File: android/kamome/src/main/java/jp/hituzi/kamome/Messenger.java
package jp.hituzi.kamome;

import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.webkit.ValueCallback;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

final class Messenger {
    @NonNull
    private static final String jsObj = "window.KM";

    public static void completeMessage(@NonNull final WebView webView, @Nullable final Object data, @NonNull final String requestId) {
        final String id = escapeForJSStringLiteral(requestId);
        if (data != null) {
            runJavaScript(String.format("%s.onComplete(%s, '%s')", jsObj, data, id), webView);
        } else {
            runJavaScript(String.format("%s.onComplete(null, '%s')", jsObj, id), webView);
        }
    }

    public static void failMessage(@NonNull final WebView webView, @Nullable final String error, @NonNull final String requestId) {
        final String id = escapeForJSStringLiteral(requestId);
        if (error != null) {
            String errMsg;
            try {
                errMsg = URLEncoder.encode(error, "utf-8");
                // The URLEncoder converts spaces to '+' and
                // the `decodeURIComponent` function on JS decodes '%20' to spaces.
                errMsg = errMsg.replaceAll("\\+", "%20");
            } catch (UnsupportedEncodingException e) {
                errMsg = error;
            }
            runJavaScript(String.format("%s.onError('%s', '%s')", jsObj, errMsg, id), webView);
        } else {
            runJavaScript(String.format("%s.onError(null, '%s')", jsObj, id), webView);
        }
    }

    public static void sendRequest(@NonNull final WebView webView, @NonNull final Request request) {
        final String name = escapeForJSStringLiteral(request.name);
        final String callbackId = escapeForJSStringLiteral(request.callbackId);
        if (request.data != null) {
            runJavaScript(String.format("%s.onReceive('%s', %s, '%s')",
                    jsObj, name, request.data, callbackId), webView);
        } else {
            runJavaScript(String.format("%s.onReceive('%s', null, '%s')",
                    jsObj, name, callbackId), webView);
        }
    }

    /**
     * Escapes a string for safe interpolation inside a single-quoted JavaScript string literal.
     */
    @NonNull
    static String escapeForJSStringLiteral(@NonNull final String value) {
        final StringBuilder sb = new StringBuilder(value.length());
        for (int i = 0; i < value.length(); i++) {
            final char c = value.charAt(i);
            switch (c) {
                case '\\': sb.append("\\\\"); break;
                case '\'': sb.append("\\'"); break;
                case '"': sb.append("\\\""); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                case '\u2028': sb.append("\\u2028"); break;
                case '\u2029': sb.append("\\u2029"); break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        return sb.toString();
    }

    private static void runJavaScript(@NonNull final String js, @NonNull final WebView webView) {
        new Handler(Looper.getMainLooper()).post(() -> {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                webView.evaluateJavascript(js, value -> {
                    // Nothing to do.
                });
            } else {
                webView.loadUrl("javascript:" + js);
            }
        });
    }
}

File: ios/kamome-framework/src/Messenger.swift
//
// Copyright (c) 2024 Hituzi Ando. All rights reserved.
//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Foundation
import WebKit

class Messenger {
    private static let jsObj = "window.KM"

    static func completeMessage(with webView: WKWebView, data: Any?, for requestID: String) throws {
        let id = escapeForJSStringLiteral(requestID)
        if let data {
            if !JSONSerialization.isValidJSONObject(data) {
                throw KamomeError.invalidJSONObject
            }

            let params = String(data: try JSONSerialization.data(withJSONObject: data), encoding: .utf8)!
            run(javaScript: "\(jsObj).onComplete(\(params), '\(id)')", with: webView)
        }
        else {
            run(javaScript: "\(jsObj).onComplete(null, '\(id)')", with: webView)
        }
    }

    static func failMessage(with webView: WKWebView, error: String?, for requestID: String) {
        let id = escapeForJSStringLiteral(requestID)
        if let error {
            let allowed = NSCharacterSet.alphanumerics.union(.init(charactersIn: "-._~"))
            let errMsg = error.addingPercentEncoding(withAllowedCharacters: allowed) ?? error
            run(javaScript: "\(jsObj).onError('\(errMsg)', '\(id)')", with: webView)
        }
        else {
            run(javaScript: "\(jsObj).onError(null, '\(id)')", with: webView)
        }
    }

    static func sendRequest(_ request: Request, with webView: WKWebView) throws {
        let name = escapeForJSStringLiteral(request.name)
        let callbackID = escapeForJSStringLiteral(request.callbackID)
        if let data = request.data {
            if !JSONSerialization.isValidJSONObject(data) {
                throw KamomeError.invalidJSONObject
            }

            let params = String(data: try JSONSerialization.data(withJSONObject: data), encoding: .utf8)!
            run(javaScript: "\(jsObj).onReceive('\(name)', \(params), '\(callbackID)')",
                    with: webView)
        }
        else {
            run(javaScript: "\(jsObj).onReceive('\(name)', null, '\(callbackID)')",
                    with: webView)
        }
    }

    /// Escapes a string for safe interpolation inside a single-quoted JavaScript string literal.
    static func escapeForJSStringLiteral(_ value: String) -> String {
        var result = ""
        result.reserveCapacity(value.count)
        for scalar in value.unicodeScalars {
            switch scalar {
            case "\\": result += "\\\\"
            case "'": result += "\\'"
            case "\"": result += "\\\""
            case "\n": result += "\\n"
            case "\r": result += "\\r"
            case "\t": result += "\\t"
            case "\u{2028}": result += "\\u2028"
            case "\u{2029}": result += "\\u2029"
            default:
                if scalar.value < 0x20 {
                    result += String(format: "\\u%04x", scalar.value)
                } else {
                    result.unicodeScalars.append(scalar)
                }
            }
        }
        return result
    }
}

private extension Messenger {
    static func run(javaScript: String, with webView: WKWebView) {
        DispatchQueue.main.async {
            webView.evaluateJavaScript(javaScript) { value, error in
                if let error = error {
                    print("[Kamome] ERROR: \(error)")
                }
            }
        }
    }
}
